Nav: Fix Styles bugs related to CommandButton entries

cicciodm opened the pull request:
#### Pull request checklist

- [x] Addresses an existing issue: Fixes [#2626](https://github.com/OfficeDev/office-ui-fabric-react/issues/2626)
- [x] Include a change request file using `$ npm run change`

#### Description of changes

When a `INavLink` has an empty URL and an `onClick` handler, a `CommandButton` element is rendered as opposed to a simple `a`. This pull request aims to fix several bugs related to styling these `CommandButton` elements.

 - "Selected" & `:hover` styles are now applied to `button` elements with `compositeLink` class
 - Button elements have now `width: 100%` under `compisiteLink` class
 - Correct indentation is applied to `CommandButton` components for different nesting levels
 - No `iconProps` are passed to `CommandButton` if none are present on the `link` object (avoids unnecessary margin)
 - The `span` for `CommandButton` description is given 0 margin when no icon is present (avoids unnecessary margin)

#### Focus areas to test

In `Nav.BasicExample.tsx`, make all `url` prop empty, and add `onClick` handlers. Then change the selected key, and observe that correct padding/correct styling is applied for all buttons.

EDIT: Some wording, removed message about change file

msftclas commented:
@cicciodm,
Thanks for having already signed the Contribution License Agreement. Your agreement has not been validated yet. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

cicciodm commented:
Running `npm run change` in the forked repo returned `No Change file is needed.`

micahgodbolt commented:
@cicciodm npm run change compares your current branch to the master branch.....which you were already on :)

To compare it to the upstream non forked repo, you can use the actual rush command `rush change -b upstream` where upstream is the name of the non fork remote.

micahgodbolt commented:
Also, can you drop in some screenshots explaining the changes you've made?

cicciodm commented:
These fixes apply to `INavLink` items where the `url` is empty and a `onClick` callback is provided, e.g.:
```
{
    name: "Documents",
    url: "",
    onClick: this._onClickHandler2,
    key: "key3"
}
```

Fix contents | Before | After
-------------|--------|------
Selected and hoverStyles applied to button | ![image](https://user-images.githubusercontent.com/3417264/30108616-3c893b8e-9303-11e7-84bd-0f497b30f428.png) | ![image](https://user-images.githubusercontent.com/3417264/30108641-5a058276-9303-11e7-95a5-25cd5e017a47.png)
width 100% for button items | ![image](https://user-images.githubusercontent.com/3417264/30111067-944b4cf0-930c-11e7-876b-ef460c144ace.png) | ![image](https://user-images.githubusercontent.com/3417264/30111083-a3ebea5c-930c-11e7-9a26-20c668597b19.png)
Correct indentation based on nesting level | ![image](https://user-images.githubusercontent.com/3417264/30111239-38c028a0-930d-11e7-9bfb-0f7ee0c83421.png) | ![image](https://user-images.githubusercontent.com/3417264/30111157-f5c3bec2-930c-11e7-8dbc-def8a80b6b29.png)
No iconProps when none are defined | ![image](https://user-images.githubusercontent.com/3417264/30111307-6c3c6fd6-930d-11e7-9192-39097daa804a.png) | ![image](https://user-images.githubusercontent.com/3417264/30111380-a7561a4a-930d-11e7-81bc-42a807e012ee.png)
0 margin span when no icon is present  | ![image](https://user-images.githubusercontent.com/3417264/30111423-d91fbe6e-930d-11e7-933e-4f91578c613f.png) | ![image](https://user-images.githubusercontent.com/3417264/30111380-a7561a4a-930d-11e7-81bc-42a807e012ee.png)

micahgodbolt commented:
A+ screencap work

micahgodbolt commented on packages/office-ui-fabric-react/src/components/Nav/Nav.tsx:
```diff
@@ -158,16 +158,22 @@ export class Nav extends BaseComponent<INavProps, INavState> implements INav {
+          this._hasExpandButton && 'isOnExpanded ' + styles.linkIsOnExpanded,
+          !link.icon && styles.commandButtonNoIcon
         ) }
+        style={ { [isRtl ? 'paddingRight' : 'paddingLeft']: paddingBefore + 'px'} }
```
style tag is soon going to be a linting error. we're moving towards using mergeStyles for all CSS declared in JS. Here's an example.

https://github.com/OfficeDev/office-ui-fabric-react/blob/9a5a9e2f3583ffda8e60a155f96c800e28a9af47/packages/experiments/src/components/LayoutGroup/LayoutGroup.tsx#L30

It can take a list of class names, or CSS objects.

cicciodm replied:
I have tried to follow your suggestion, by changing the style as follows:
```
<CommandButton
  className={
    mergeStyles(
      'ms-Nav-link ms-Nav-linkButton',
      styles.link,
      this._hasExpandButton && 'isOnExpanded ' + styles.linkIsOnExpanded,
      !link.icon && styles.commandButtonNoIcon,
      isRtl && {
        paddingRight: paddingBeforePixels
      },
      !isRtl && {
        paddingLeft: paddingBeforePixels
      }
    ) as string
  }
. . .
```
However, it looks like the `ms-Nav-Link` class already has a definition for `padding-left`: this definition was overridden by the `styles` prop passed to `CommandButton`, but passing the CSS object to mergeStyles does not have the same effect.

Do you have any thoughts on how to fix this?

( appending `!important` to `paddingBeforePixels` fixes the issue, but I don't feel that's the right solution)

micahgodbolt replied:
Can you move the padding from ms-Nav-Link into this mergeStyles block? Under what circumstances is that padding-left needed?

cicciodm replied:
Looking more carefully, I noticed that the `.link` class was responsible for that padding.

I have removed all padding definitions from the CSS, and let the TS code + `mergeStyles` handle it now. I still had to add an `!important` for the `CommandButton` types, because some internal styles were overriding the styles passed in.

cicciodm commented:
The problem is that the style that's being overridden is a style belonging to all `CommandButton` Fabric elements. 

While I am passing a merged style as a `className` to the `CommandButton`, the order in which the CSS styles are loaded makes it such the `CommandBar` stylesheet takes precedence over the style I have defined.

![image](https://user-images.githubusercontent.com/3417264/30220972-df5cd2d4-94c1-11e7-8a7b-eb7a3ca671b9.png)

From the screenshot, you can see the Red style is the one the was defined in `Nav`, which is being overridden by the blue definition. If I hover over the `.css-h9c8km` class, this is what's highlighted on the page:

![image](https://user-images.githubusercontent.com/3417264/30221053-279f31f4-94c2-11e7-971e-2103b0faf525.png)

I am not sure how to move that padding definition to a mergeStyles, since it belongs to another component!

micahgodbolt commented:
@cicciodm do you have access to the rendering of the button? If so, you can pass in a custom style object to remove that padding. This is not specificity based, so order won't matter. It changes the value before render takes place. 

```
<CommandButton styles={{root: {padding: 0}} />
```

cicciodm commented:
@micahgodbolt The original code *does have* a `style` object passed in to the `CommandButton` component, but the reason why I made these CSS changes is because of your comment, a couple of iterations ago, that the `style` prop will soon be a linting error, and that we should move to using `mergeStyles` to create a `className` prop. 

Does this still apply? Moving the CSS object to `styles` would definitely simplify things.

micahgodbolt commented:
Sorry about that! To clarify, the "style" prop (i.e. inline style) will be a linting error. The button "styles" prop is react prop for passing an IButtonStyles object.

I'm sorry that can be a bit confusing :)

cicciodm commented:
Thanks for the reply, and apologies for the late response.

I have changed the code to use the `styles` property of the `CommandButton` component as you requested.

I also fixed a small bug in how the position of the chevron button is set in case of composite links: previously, a `margin` was applied to the `button` component, but the style was not correctly applied. This would be evident in nested groups.

I am now calculating the correct `left/right` position, and applying it through `mergeStyles` rather than CSS.

Before | After
-------|------
![image](https://user-images.githubusercontent.com/3417264/30523301-8bc750ca-9bdf-11e7-8f04-f0457f4076ce.png) | ![image](https://user-images.githubusercontent.com/3417264/30523302-930ff346-9bdf-11e7-9f8c-177e254cf4f2.png)

micahgodbolt commented:
Sorry about the delay on review. I'm pulling this down to take a look at it.

dzearing commented on packages/office-ui-fabric-react/src/components/Nav/Nav.scss:
```diff
@@ -103,13 +103,16 @@ $linkRightPadding: 20px;
   display: block;
   position: relative;
 
+  & button {
```
please remove this rule. It is a huge assumption to assume all buttons in this element should fill.

micahgodbolt commented:
I think we should wait for #2916 before merging this PR. It'll allow the styles in this PR to be simplified quite a bit.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/OfficeDev/office-ui-fabric-react?pullRequest=2728) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/OfficeDev/office-ui-fabric-react?pullRequest=2728) before we can accept your contribution.<br/><br/>:x: cicciodm [sign now](https://cla.opensource.microsoft.com/OfficeDev/office-ui-fabric-react?pullRequest=2728)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/OfficeDev/office-ui-fabric-react?pullRequest=2728) it.</sub>

cicciodm commented:
I have made quite a significant change for this, which might actually require some design input and maybe a different pull request entirely.

As @micahgodbolt pointed out in a separate conversation, there is no reason for the functions `_renderAnchorLink` and `_renderButtonLink` to be separate: the existing `ActionButton` component will render just fine links with an url as `a` elements, without losing accessibility. I have therefore unified the 2 codepaths into a single call to `_renderNavLink`.

At this point, everything is rendered as a `Button`, and every `Button` has a placeHolder icon, [(#2916)](https://github.com/OfficeDev/office-ui-fabric-react/pull/2916) there's no need for complex calculations of a left padding based on whether a expand/collapse button is present.

Since everything is rendered as a button, the alignment of the icons will look different, and now all button labels will be aligned, whether an icon is present or not. This should probably be discussed with design.

![image](https://user-images.githubusercontent.com/3417264/31349129-39ed4c56-ad22-11e7-9a1b-9a9e950be92c.png)

Please advise on how to best review this change: should I open a new, cleaner PR?

cicciodm commented:
Also, I was told that internal collaborators don't need to sign a CLA, but even if I did now the system is asking for me to sign?

cicciodm commented:
Also, I am very open to suggestions on how to get a button to have text ellipsis without passing so many overrides to `buttonStyles`:
```
textContainer: {
        overflow: "hidden",
      },
      label: {
        whiteSpace: "nowrap",
        textOverflow: "ellipsis",
        overflow: "hidden",
        lineHeight: "36px"
      }
```

dzearing commented on packages/office-ui-fabric-react/src/components/Nav/Nav.scss:
```diff
@@ -103,13 +103,16 @@ $linkRightPadding: 20px;
   display: block;
   position: relative;
 
+  & button {
```
This is way too generic. Why can't we target a specific class?

Think of it this way; if a nav element nests multiple buttons, what happens?

cicciodm replied:
We definitely can now that every Nav entry is a button

dzearing commented on packages/office-ui-fabric-react/src/components/Nav/Nav.scss:
```diff
@@ -142,6 +143,7 @@ $linkRightPadding: 20px;
 
   &.compositeLinkIsSelected {
     & a,
```
any button or any anchor is way too specific. What about immediate descendant?

dzearing replied:
or better, a specific class target.

cicciodm replied:
I should be able to assing these properties to the `link` class now, which is applied to all `NavLinks`

dzearing commented:
@cicciodm Sorry for this to take so long. Part of the issue is that we don't have enough eyes on nav specific problem, part of it is that we do not want to break partners with assumptions that we make here. So things like css rules that target "any button" make a lot of people nervous and unsure about whether it's a good change or not. We just don't want to fix one thing and then cause a bunch of regressions.

dzearing commented:
@inateeg Can you please take a look here? Because I think some of the css rules may affect your Nav scenarios.

cicciodm commented:
@dzearing I have addressed your comments, and removed the generic CSS selectors as mentioned. Always available to make more changes.

dzearing commented:
This looks right to me!

cicciodm commented:
Thanks for merging!

micahgodbolt commented:
Thanks for all your work!
